### ISSUE TITLE
Market Maker Incentivization Worker Proposal

### ISSUE THREAD

**Issue by bytemaster** [cryptonomex/graphene#475]
## Introduction

The success and failure of a market depends to a large degree on the number of participants in that market. The number of participants depends upon the liquidity in a market.  This is a chicken and egg problem that must be resolved by bootstrapping a new market with liquidity.  This bootstrapping process is very risky because it involves trading in an illiquid market. With the proper incentives it can become profitable to bootstrap a market until no additional incentives are necessary.
## Market Makers

A market maker is anyone who has their order filled after being left open on the books for a minimum period of time (at least 1 block). These individuals leave an open order ready to meet the demand for someone looking to buy and/or sell.  Leaving orders on the books that are never filled does not qualify for being a maker even if they add the appearance of depth. To be more specific, a market maker buys and resells over and over again and pockets the spread between buying and selling as their profit.  This is not a risk free operation if the market starts to move consistently in any one direction.
## Flexible Maker Fees 

The first step toward increasing liquidity is to reduce the artificial costs associated with trading in a market. This means simply lowering the trading fees to 0 for those who provide liquidity (makers) while only charging those who consume liquidity (takers).  Under this proposal we would introduce a new parameter to assets that allows the issuer to configure different market fees for makers vs takers.  Early on it may be desirable to have 0 maker fee, but for the sake of flexibility we will parameterize it.
## Rewarding Market Makers

Simply reducing the fees does little to reduce the risks (costs) associated with trading in an illiquid market. These costs are disproportionately borne by the early adopters of a market, especially in volatile markets. In order to offset these costs these market makers should be compensated for their risk in a manner that does not impose costs to participants outside the market.  The goal is to reallocate the income generated by the market fees charged to the takers, to the makers.   

An instantaneous reallocation does little more than shift the equilibrium price. Instead we propose to redistribute future market fees generated after a market has been bootstrapped to the risk takers that facilitated the growth of that market. Each market is like a startup business and like all startup businesses, some capital is required get started. In the case of a market, the startup capital required is liquidity provision. Shares in a company are usually distributed proportional to how much capital each founder contributed. As the company matures and raises future rounds of capital the company can sell their shares for higher and higher prices (offer less shares for the same capital).   
## Definitions

_MSHARES_ - a special asset that has maximum supply M and is rewarded to users based upon the percent of the daily liquidity (volume) they provide to a particular ASSET. Every ASSET has the potential to have their own unique MSHARES. 
_RESERVE_ - The total remaining MSHARES to be distributed over time
_BUCKET_ - a collection of MSHARES that is available to be divided among current liquidity providers.  This is used to implement the [leaky bucket algorithm](https://en.wikipedia.org/wiki/Leaky_bucket)
_DECAY_RATE_ - The maximum percent of the RESERVE that can be moved into the BUCKET per second.  This defines the half life of the reserve and the extent to which early liquidity providers have an advantage over later liquidity providers.

_B_ - the total number of MSHARES in the BUCKET at any given point in time.
_Bmax_ - the maximum MSHARES the BUCKET may accumulate (a multiple of the DECAY_RATE).  
_Vavg_ - An approximation to the average volume in the market
_Vperiod_ - the number of seconds to average volume over (1 week recommended)
_V_ - the volume of the maker order being filled (measured in UIA units)
_REWARD_ - the reward in MSHARES that a MAKER receives 
_T_  = seconds since last match
_Tmin_ = the minimum time a order must be on the books to be considered for maker rewards.
## Calculations Performed on Match

Top off the bucket (but cap at `Bmax`):

```
    Bchange = min( B + RESERVE * DECAY_RATE * T, Bmax ) - B
    RESERVE = RESERVE - Bchange
    B = B + Bchange
```

Reward a fraction of the bucket:

```
    REWARD = B * V / (Vavg + V)
```

Remove Reward from Bucket:

```
    B  = B - REWARD
```

Update Weighted Average of Volume

```
    Vavg  = Vavg * MAX(0,Vperiod-T)/Vperiod + V
```

After calculating the REWARD the Maker's balance of MSHARES is increased.
## Distributing Market Fees

Every maintenance interval the accumulated market fees for an UIA are used to purchase MSHARES in the UIA/MSHARES market. The percentage of market fees that are directed to repurchase MSHARES is a parameters that is set by the issuer. It may be increased, but never decreased.  

All parameters are configured by the asset issuer including DECAY_RATE, Bmax, RESERVE, Vperiod, Tmin, maker fee, and percent of market fees allocated to the maker. An issuer may revoke their permission to change the parameters and thereby reduce the risk of the issuer changing the terms and impacting the value of MSHARES.
## Maker Issued Asset

The result of this proposal is the creation of a new type of asset called the _Maker Issued Asset_ that can only be created/destroyed by the algorithm above. 
## Funding this Proposal

It is our belief that the cost of implementing a feature must be less than the present value of a feature. It is the belief of Cryptonomex that this particular feature is worth much more than the cost to implement it and therefore Cryptonomex will be implementing it speculatively in exchange for a cut of all MSHARES for all assets that use this feature to improve their liquidity.   

We will create a new NETWORK ISSUED ASSET, called MAKER that will automatically be repurchased with 20% of all MAKER ISSUED ASSETS ever created. This asset with have a maximum supply of 1,000,000 MAKER and initially be 100% allocated to Cryptonomex in exchange for implementing this feature.  Cryptonomex will sell MAKER into the market to cover the cost of development. 
## Approving this Proposal

In order for this proposal to be approved by the network a worker proposal will be created that will pay 100,000 BTS to Cryptonomex. If the proposal gets funded then it will be assumed that the required hardfork has been approved by stakeholders and development will compense. 
## Understanding by Example

If a particular asset issuer defines their DECAY rate such that 50% of all MSHARES are allocated in the first year, then someone who provides 100% of the liquidity would earn 50% of all future market fees in exchange for liquidity provision.  It does not matter what the volume is over that year, all that matters is the relative percentage of liquidity provided by every market maker.   

If the market starts out small with $10K / 24hr volume, and then turns into a $30M 24/hr volume then the market fees of 0.2% will yield $30,000 per day, half of which would belong to the individual who provided a meer $10K of liquidity in the first year.   Meanwhile those who provide $10K of liquidity once $30M of volume is reached will get a very small number of MSHARES.   

Assuming a market provides a useful function for traders, these incentives should easily encourage early adopters to bootstrap the market. 
## Preventing Abuse

All systems must be analysed for potential abuse. In this case we are concerned about a single bad-actor attempting to create fake volume in order to profit without actually providing liquidity.  Abuse is prevented by three factors:
1. There is a non-zero cost to creating fake volume in the form of order creation fees charged by the network.
2. Creating fake volume requires also being a "taker" and paying the taker's market fee.  The asset issuer will only allocate a fraction of the taker fees to repurchase.  This means that there is a non-zero cost of creating MSHARES.
3. In the event that it is profitable to create MSHARES through fake volume, there will be market competition to fight over the "free money".  This would work like mining difficulty as many different individuals compete to create "fake volume" until percent of MSHARES each market participants earns approaches the cost of creating those MSHARES.

Our conclusion is that the design of this algorithm is resistant to abuse. If abuse is attempted the network and the issuer should profit greatly.

**Comment by Krolling** [comment]
## Question about what fees go to purchase MSHARES

Do 100% of collected fees go to purchase MSHARES? (The example makes it appear so)
Where is the profit for BitShares if we're giving away all the fees?
How will this affect the original model of 20% of fees going to the network and 80% to the referral system?

## Question about fees and referrals

If fees are going to go to repurchase MSHARES, will that hurt referral income?

## Question about fee pool/collected fees

When fees are paid in UIA (or anything not BTS), which UIA/BTS/BTA is used to repurchase MSHARES? If in BTS, that will drain the fee pool and require more maintenance.

## Question about "Funding this Proposal"

I'm a little unclear about the details of "Funding this Proposal"

"Calculations Performed on Match" says, "after calculating the REWARD the Maker's balance of MSHARES is increased." And later on, under "Funding this Proposal" it says, "We will create a new NETWORK ISSUED ASSET, called MAKER that will automatically be repurchased with 20% of all MAKER ISSUED ASSETS ever created."

So does that mean:
1) A trade occurs in BTS:USD, and according to the algorithm in "Calculations Performed on Match", the reward should be 100 MSHARES.USD
2) 80 MSHARES.USD goes to the Maker
3) 20 MSHARES.USD go to purchase as many MAKER as it can on the MAKER:MSHARES.USD market?

What happens if there are no asks on the MAKER:MSHARES.USD market? (i.e., it is a new Maker Issued Asset Market?)

Game theory says that cryptonomex should sell 1 MAKER for infinity MSHARES.USD:MAKER. This is because cryptonomex has all the MAKERs so there is no market. The outcome is that cryptonomex keeps 20% of all MSHARES.USD and all their MAKER tokens. That's fine, just wanted to point that out. Why not hardcode it that 20% goes directly to cryptonomex? Is it because you want the option to sell MAKER earlier on the market (before MSHARES are created) get a faster return?

EDIT:

# Example can't be true

> If a particular asset issuer defines their DECAY rate such that 50% of all MSHARES are allocated in the first year, then someone who provides 100% of the liquidity would earn 50% of all future market fees in exchange for liquidity provision.

That means that cryptonomex got 0 MSHARES. It should be:

> If a particular asset issuer defines their DECAY rate such that 50% of all MSHARES are allocated in the first year, then someone who provides 100% of the liquidity would earn 40% of all future market fees in exchange for liquidity provision and cryptonomex will receive 10% of all future market fees.

**Comment by bytemaster** [comment]
## Updating Example

I have updated the example to state that MAKER owners would receive 10% rather than saying that Cryptonomex will receive 10% of all future market fees.  

## Fees that go to purchase MSHARES

When there is a market fee it is denominated in UIA and is value is a percent of the the UIA received from the trade (ie: 0.2%).  If the trade was for $1000, then $2 would be the "market fee" paid by the "taker".      This $2 will be further divided according to a parameters set by the _issuer_, the issuer could allocate 50% for issuer profits, and 50% for maker.  In this case $1 would be used to repurchased MSHARES for UIA.   The total maker rewards would be:  REWARD =  B \* 1000 / (Vavg + 1000).   Of this 20% of REWARD will be used to purchase MAKER in the MSHARES/MAKER market, and 80% will go to the user who performed the market maker role.   

BitShares would receive the ORDER_CREATION_FEE which is fixed regardless of the size of the order and is about $0.20 for regular users and $0.04 for lifetime members.   This ORDER_CREATION_FEE is distributed to the referral program / network like all other non-market fees.   

The only market fees earned by BTS holders would be the 50% (variable) allocated to the issuer when the issuer is the committee.   These funds are at the discretion of the committee.   

Bottom line, BTS holders make the same money they use to make. This proposal simply offers UIA issuers the opportunity to increase their profits by bootstrapping their liquidity with a stake in future revenue from the market.

**Comment by NaturalCoder** [comment]
If we are incentivising them then does not make sense charge a fee, the fee should be 0.1 BTS or same other anti-spam value, the same to barrow.

**Comment by vikramrajkumar** [comment]
This issue was moved to bitshares/bitshares-core#137
